storage/pebble/chunks_queue.go

```diff
@@ -49,40 +51,51 @@ func (q *ChunksQueue) Init(defaultIndex uint64) (bool, error) {
 // A true will be returned, if the locator was new.
 // A false will be returned, if the locator was duplicate.
 func (q *ChunksQueue) StoreChunkLocator(locator *chunks.Locator) (bool, error) {
-	err := operation.RetryOnConflict(q.db.Update, func(tx *badger.Txn) error {
+	q.storing.Lock()
+	defer q.storing.Unlock()
+
+	var alreadyExist bool
+	err := operation.HasChunkLocator(locator.ID(), &alreadyExist)(q.db)
+	if err != nil {
+		return false, fmt.Errorf("could not check if chunk locator exists: %w", err)
+	}
+
+	// was trying to store a duplicate locator
+	if alreadyExist {
+		return false, nil
+	}
+
+	err = operation.WithReaderBatchWriter(q.db, func(tx storage.PebbleReaderBatchWriter) error {
+		r, w := tx.ReaderWriter()
 		// make sure the chunk locator is unique
-		err := operation.InsertChunkLocator(locator)(tx)
+		err := operation.InsertChunkLocator(locator)(w)
 		if err != nil {
 			return fmt.Errorf("failed to insert chunk locator: %w", err)
 		}
 
 		// read the latest index
 		var latest uint64
-		err = operation.RetrieveJobLatestIndex(JobQueueChunksQueue, &latest)(tx)
+		err = operation.RetrieveJobLatestIndex(JobQueueChunksQueue, &latest)(r)
 		if err != nil {
 			return fmt.Errorf("failed to retrieve job index for chunk locator queue: %w", err)
 		}
 
 		// insert to the next index
 		next := latest + 1
```

Comment:
In order to prevent dirty reads of the `latest`, we have to introduce a mutex lock (`storing`) here, so that it's concurrent-safe.